src/BSON/Int64.c

```diff
@@ -244,6 +263,229 @@ static int php_phongo_int64_compare_objects(zval* o1, zval* o2)
+ * squaring algorithm. This is necessary because in case the result exceeds
+ * the range of a int64_t, we want PHP to return a float as it would when
+ * using 64-bit values directly. We can't use anything involving zend_long
+ * here as this would limit us to 32 bits on a 32-bit platform. This also
+ * prohibits us from falling back to PHP's default functions after unwrapping
+ * the int64_t from the php_phongo_int64_t instance. */
+static int64_t phongo_pow_int64(int64_t base, int64_t exp)
+{
+	if (exp == 0) {
+		return 1;
+	}
+
+	if (exp % 2) {
+		return base * phongo_pow_int64(base * base, (exp - 1) / 2);
+	}
+
+	return phongo_pow_int64(base * base, exp / 2);
+}
+
+#define OPERATION_RESULT_INT64(zval, value)                                          \
+	if (Z_TYPE_P((zval)) == IS_OBJECT && Z_OBJCE_P((zval)) == php_phongo_int64_ce) { \
+		Z_INT64_OBJ_P((zval))->integer = (value);                                    \
+	} else {                                                                         \
+		ZVAL_INT64_OBJ((zval), value);                                               \
+	}
+
+#define PHONGO_GET_INT64(int64, zval)                                                       \
+	if (Z_TYPE_P((zval)) == IS_LONG) {                                                      \
+		(int64) = Z_LVAL_P((zval));                                                         \
+	} else if (Z_TYPE_P((zval)) == IS_OBJECT && Z_OBJCE_P((zval)) == php_phongo_int64_ce) { \
+		(int64) = Z_INT64_OBJ_P((zval))->integer;                                           \
+	} else {                                                                                \
+		return FAILURE;                                                                     \
+	}
+
+/* Overload arithmetic operators for computation on int64_t values.
+ * This ensures that any computation involving at least one php_phongo_int64_t
+ * results in a php_phongo_int64_t value, regardless of whether the result
+ * would fit in an int32_t or not.
+ * Note that ZEND_(PRE|POST)_(INC|DEC) are not handled here: when checking for
+ * a do_operation handler for inc/dec, PHP calls the handler with a ZEND_ADD
+ * or ZEND_SUB opcode and the same pointer for result and op1, and a ZVAL_LONG
+ * of 1 for op2. This means that we can neither specially handle inc/dec
+ * operators, nor can we ensure correct behaviour for pre/post operators,
+ * leading to a difference in behaviour compared to int values. */
+static zend_result php_phongo_int64_do_operation(zend_uchar opcode, zval* result, zval* op1, zval* op2)
+{
+	int64_t value1, value2, lresult;
+
+	PHONGO_GET_INT64(value1, op1);
+
+	switch (opcode) {
+		case ZEND_ADD:
+			PHONGO_GET_INT64(value2, op2);
+
+			lresult = value1 + value2;
+
+			// Check for overflow by comparing the sign bits of the operands and the result
+			if ((value1 & INT64_MIN) == (value2 & INT64_MIN) && (value1 & INT64_MIN) != (lresult & INT64_MIN)) {
```

Comment:
I assume this was adapted from the [`#else` block of `fast_long_add_function()`](https://github.com/php/php-src/blob/php-8.2.6/Zend/zend_operators.h#L691-L703). If so, can you leave a comment for future reference?

I was initially confused as to why you were using `INT64_MIN` here. Defining an `INT64_SIGN_MASK` (or similar) constant from `INT64_MIN` may make this and the `ZEND_SUB` case below more readable.

Side note: this isn't a critical code path, but I found it interesting that php-src performs the addition twice. I think it makes sense to do it once like you've done here with `lresult`, but I'm curious if there's a particular reason php-src did not. :shrug: 

----

And if you'd like to add the relative conditions to the comment (like I suggested for `ZEND_SUB` below):

 * Adding a positive to a positive yields a negative
 * Adding a negative to a negative (i.e. subtraction) yields a positive

Reply:
Added references here and for other arithmetic operations.